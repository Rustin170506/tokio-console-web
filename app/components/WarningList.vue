<template>
    <div class="p-4 flex-1 space-y-4">
        <template v-if="warnings.length">
            <UAlert
                v-for="(warning, index) in warnings"
                :key="`warning-${index}`"
                icon="i-heroicons-exclamation-triangle-20-solid"
                color="yellow"
                variant="outline"
                :title="warning.title"
                :description="warning.description"
            />
        </template>
        <div v-else class="flex flex-col justify-center items-center h-full">
            <div class="flex flex-col items-center justify-center p-4">
                <UIcon
                    name="i-heroicons-check-circle"
                    class="w-16 h-16 text-green-500"
                />
                <h3 class="mt-2 text-lg font-medium text-gray-900">
                    No Warnings
                </h3>
            </div>
        </div>
    </div>
</template>

<script setup lang="ts">
import { toTaskWarningItems, type WarningItem } from "~/types/warningItem";

const { tasksData, lastUpdatedAt } = useTasks();
const warnings = computed(() => {
    const lastUpdated = lastUpdatedAt.value;
    if (!lastUpdated) {
        return [];
    }
    return Array.from(tasksData.value.values()).reduce((acc, task) => {
        const taskWarningItems = toTaskWarningItems(task.warnings);
        if (taskWarningItems) {
            acc.push(...taskWarningItems);
        }
        return acc;
    }, [] as WarningItem[]);
});
</script>
