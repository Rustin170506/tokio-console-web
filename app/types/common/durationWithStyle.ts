import type { Duration } from "./duration";

/**
 * DurationWithStyle is a wrapper around the Duration type that includes a tailwind class to use for this duration.
 * It is used to display the duration in a color that represents the magnitude of the duration.
 */
export class DurationWithStyle {
    value: Duration;
    // The tailwind class to use for this duration.
    class: string;

    constructor(value: Duration, className: string) {
        this.value = value;
        this.class = className;
    }

    /**
     * Provides a human-readable string representation of the duration.
     * @returns a string representation of the duration.
     */
    toString(): string {
        return this.value.toString();
    }

    /**
     * Returns the duration as a number of microseconds.
     * Used for sorting.
     * @returns the duration as a number of microseconds.
     */
    valueOf(): number {
        return this.value.valueOf();
    }
}

export function getDurationWithClass(duration: Duration): DurationWithStyle {
    const days = duration.asDays();
    const hours = duration.asHours();
    const minutes = duration.asMinutes();
    const seconds = duration.asSeconds();
    const milliseconds = duration.asMilliseconds();

    let className: string;

    if (days >= 1) {
        className = "text-blue-500 dark:text-blue-300";
    } else if (hours >= 1) {
        className = "text-cyan-500 dark:text-cyan-300";
    } else if (minutes >= 1) {
        className = "text-green-500 dark:text-green-300";
    } else if (seconds >= 1) {
        className = "text-yellow-500 dark:text-yellow-300";
    } else if (milliseconds >= 1) {
        className = "text-red-500 dark:text-red-300";
    } else {
        className = "text-gray-500 dark:text-gray-300";
    }

    return new DurationWithStyle(duration, className);
}
