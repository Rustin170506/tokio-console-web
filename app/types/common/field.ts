import type { Metadata } from "./metadata";
import { Field as ProtoField } from "~/gen/common_pb";
import type { Location } from "~/gen/common_pb";

export function truncateRegistryPath(s: string): string {
    const regex = /.*\/cargo(\/registry\/src\/[^/]*\/|\/git\/checkouts\/)/;
    return s.replace(regex, "<cargo>/");
}

export function formatLocation(loc?: Location): string {
    if (loc) {
        let result = "";
        if (loc.modulePath) {
            result = loc.modulePath;
        } else if (loc.file) {
            const truncated = truncateRegistryPath(loc.file);
            result = truncated;
        } else {
            return "<unknown location>";
        }

        if (loc.line !== undefined) {
            result += `:${loc.line}`;

            if (loc.column !== undefined) {
                result += `:${loc.column}`;
            }
        }

        return result;
    }
    return "<unknown location>";
}

export enum FieldValueType {
    Bool = "Bool",
    Str = "Str",
    U64 = "U64",
    I64 = "I64",
    Debug = "Debug",
}

export class FieldValue {
    type: FieldValueType;
    value: boolean | string | bigint;

    constructor(type: FieldValueType, value: boolean | string | bigint) {
        this.type = type;
        this.value = value;
    }

    truncateRegistryPath(): FieldValue {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            typeof this.value === "string"
        ) {
            this.value = truncateRegistryPath(this.value);
        }
        return this;
    }

    ensureNonempty(): FieldValue | undefined {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            this.value === ""
        ) {
            return undefined;
        }
        return this;
    }

    static fromProto(field: ProtoField): FieldValue {
        switch (field.value.case) {
            case "debugVal":
                return new FieldValue(FieldValueType.Debug, field.value.value);
            case "strVal":
                return new FieldValue(FieldValueType.Str, field.value.value);
            case "u64Val":
                return new FieldValue(FieldValueType.U64, field.value.value);
            case "i64Val":
                return new FieldValue(FieldValueType.I64, field.value.value);
            case "boolVal":
                return new FieldValue(FieldValueType.Bool, field.value.value);
            default:
                throw new Error("Invalid field value case");
        }
    }
}

export class Field {
    static SPAWN_LOCATION = "spawn.location";
    static KIND = "kind";
    static NAME = "task.name";
    static TASK_ID = "task.id";

    name: string;
    value: FieldValue;

    constructor(name: string, value: FieldValue) {
        this.name = name;
        this.value = value;
    }

    static fromProto(field: ProtoField, meta: Metadata): Field | undefined {
        const nameValue = field.name;
        if (nameValue.case === undefined) {
            return undefined;
        }

        let name = "";
        switch (nameValue.case) {
            case "strName": {
                name = nameValue.value;
                break;
            }
            case "nameIdx": {
                const index = nameValue.value;
                const metaId = field.metadataId?.id;
                if (meta.id !== metaId) {
                    return undefined;
                }
                if (meta.fieldNames[Number(index)]) {
                    name = meta.fieldNames[Number(index)];
                } else {
                    return undefined;
                }
                break;
            }
        }

        let value = FieldValue.fromProto(field).ensureNonempty();
        // If the field is invalid or it has a string value which is empty, return null
        if (!name || !value) {
            return undefined;
        }

        // If the field name is 'SPAWN_LOCATION', truncate the registry path
        if (name === Field.SPAWN_LOCATION) {
            value = value.truncateRegistryPath();
        }

        return new Field(name.toString(), value);
    }
}
