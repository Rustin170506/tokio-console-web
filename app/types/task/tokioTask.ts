import { Duration, Timestamp } from "../common/duration";
import type { Field } from "../common/field";
import { Warning, type Linter, type Warn } from "../warning/warn";
import type { TokioTaskStats } from "./tokioTaskStats";

export enum TaskState {
    Completed,
    Running,
    Scheduled,
    Idle,
}

export enum TaskLintResult {
    Linted,
    RequiresRecheck,
}

export class TokioTask {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tokio::task::Id` in the remote tokio runtime.
    taskId?: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    // A precomputed short description string used in the async ops table
    shortDesc: string;
    // Fields that don't have their own column, pre-formatted
    fields: Array<Field>;
    // The task statistics that are updated over the lifetime of the task.
    stats: TokioTaskStats;
    // The target of the span representing the task.
    target: string;
    // The name of the task
    name?: string;
    // The source file and line number the task was spawned from
    location: string;
    // The kind of task, currently one of task, blocking, block_on, local
    kind: string;
    // Currently active warnings for this task.
    /* eslint-disable no-use-before-define */
    warnings: Array<Warn<TokioTask>>;

    constructor(
        id: bigint,
        taskId: bigint | undefined,
        spanId: bigint,
        shortDesc: string,
        formattedFields: Array<Field>,
        stats: TokioTaskStats,
        target: string,
        name: string | undefined,
        location: string,
        kind: string,
    ) {
        this.id = id;
        this.taskId = taskId;
        this.spanId = spanId;
        this.shortDesc = shortDesc;
        this.fields = formattedFields;
        this.stats = stats;
        this.target = target;
        this.name = name;
        this.location = location;
        this.kind = kind;
        this.warnings = [];
    }

    totalDuration(since: Timestamp): Duration {
        const duration = since.subtract(this.stats.createdAt);
        return this.stats.total ?? duration;
    }

    busyDuration(since: Timestamp): Duration {
        if (this.stats.lastPollStarted) {
            if (
                !this.stats.lastPollEnded ||
                this.stats.lastPollStarted > this.stats.lastPollEnded
            ) {
                // In this case the task is being polled at the moment.
                const currentTimeInPollDuration = since.subtract(
                    this.stats.lastPollStarted,
                );
                return this.stats.busy.add(currentTimeInPollDuration);
            }
        }
        return this.stats.busy;
    }

    scheduledDuration(since: Timestamp): Duration {
        if (this.stats.lastWake) {
            if (
                !this.stats.lastPollStarted ||
                this.stats.lastWake > this.stats.lastPollStarted
            ) {
                // In this case the task is scheduled, but has not yet been polled.
                const currentTimeSinceWakeDuration = since.subtract(
                    this.stats.lastWake,
                );
                return this.stats.scheduled.add(currentTimeSinceWakeDuration);
            }
        }
        return this.stats.scheduled;
    }

    idleDuration(since: Timestamp): Duration {
        if (this.stats.idle) {
            return this.stats.idle;
        } else {
            const total = this.totalDuration(since);
            const busy = this.busyDuration(since);
            const scheduled = this.scheduledDuration(since);
            if (total.greaterThan(busy.add(scheduled))) {
                return total.subtract(busy.add(scheduled));
            }
        }
        return new Duration(BigInt(0), 0);
    }

    isRunning(): boolean {
        if (this.stats.lastPollStarted && this.stats.lastPollEnded) {
            return this.stats.lastPollStarted.greaterThan(
                this.stats.lastPollEnded,
            );
        }
        if (
            this.stats.lastPollStarted &&
            this.stats.lastPollEnded === undefined
        ) {
            return true;
        }
        return false;
    }

    isScheduled(): boolean {
        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }
        return false;
    }

    isCompleted(): boolean {
        return this.stats.total !== undefined;
    }

    isBlocking(): boolean {
        return ["block_on", "blocking"].includes(this.kind);
    }

    // Returns whether this task has signaled via its waker to run again.
    // Once the task has been polled, this is changed back to false.
    isAwakened(): boolean {
        // Before the first poll, the task is waiting on the executor to run it
        // for the first time.
        if (this.totalPolls() === 0n) {
            return true;
        }

        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }

        if (this.stats.lastWake && this.stats.lastPollStarted === undefined) {
            return true;
        }

        return false;
    }

    totalPolls(): bigint {
        return this.stats.polls;
    }

    state(): TaskState {
        if (this.isCompleted()) {
            return TaskState.Completed;
        }

        if (this.isRunning()) {
            return TaskState.Running;
        }

        if (this.isScheduled()) {
            return TaskState.Scheduled;
        }

        return TaskState.Idle;
    }

    wakerCount(): bigint {
        return this.stats.wakerClones - this.stats.wakerDrops > 0n
            ? this.stats.wakerClones - this.stats.wakerDrops
            : 0n;
    }

    /**
     * Returns the elapsed time since the task was last woken, relative to
     * given `now` timestamp.
     *
     * Returns `undefined` if the task has never been woken, or if it was last woken
     * more recently than `now` (which *shouldn't* happen as long as `now` is the
     * timestamp of the last stats update...)
     */
    sinceWakeDuration(now: Timestamp): Duration | undefined {
        if (this.stats.lastWake) {
            return now.subtract(this.stats.lastWake);
        }
    }

    durationPercent(now: Timestamp, amt: Duration): number {
        const total = this.totalDuration(now);
        let percent = (amt.asSeconds() / total.asSeconds()) * 100;
        if (percent > 100) {
            percent = 100;
        }

        return percent;
    }

    lint(linters: Array<Linter<TokioTask>>): TaskLintResult {
        this.warnings.length = 0;
        let recheck = false;
        linters.forEach((linter) => {
            const result = linter.check(this);
            switch (result) {
                case Warning.Ok:
                    break;
                case Warning.Recheck:
                    recheck = true;
                    break;
                case Warning.Warn:
                    this.warnings.push(linter);
                    break;
            }
        });
        if (recheck) {
            return TaskLintResult.RequiresRecheck;
        } else {
            return TaskLintResult.Linted;
        }
    }
}
