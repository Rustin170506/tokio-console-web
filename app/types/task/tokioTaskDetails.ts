import { Duration } from "../common/duration";
import {
    deserializeHistogram,
    Percentile as RawPercentile,
    DurationCount as RawDurationCount,
} from "~/gen/histogram";
import type { TaskDetails } from "~/gen/tasks_pb";

/**
 * Percentile represents a percentile value and its corresponding duration.
 * For example, a 99th percentile of 100ms means that 99% of the data points are less than or equal to 100ms.
 */
export interface Percentile {
    percentile: number;
    duration: Duration;
}

/**
 * DurationCount represents a duration and the number of times that duration was observed.
 */
export interface DurationCount {
    duration: Duration;
    count: bigint;
}

/**
 * DurationDetails represents the details of a duration histogram.
 * It includes the percentiles, histogram, max, and min values.
 */
export interface DurationDetails {
    percentiles: Percentile[];
    histogram: DurationCount[];
    max: Duration;
    min: Duration;
}

/**
 * TokioTaskDetails represents the details of a tokio task.
 * It includes the poll times and scheduled times.
 */
export interface TokioTaskDetails {
    pollTimes: DurationDetails;
    scheduledTimes?: DurationDetails;
}

/**
 * fromProtoTaskDetails converts a protobuf TaskDetails to a TokioTaskDetails.
 * It includes the poll times and scheduled times histograms.
 * @param details - the protobuf TaskDetails
 * @param width - the width of the details chart
 * @returns the TokioTaskDetails
 */
export function fromProtoTaskDetails(
    details: TaskDetails,
    width: number,
): TokioTaskDetails {
    const toDurationPercentile = (p: RawPercentile) => ({
        percentile: p.percentile,
        duration: Duration.fromNano(p.value),
    });
    const toDurationCount = (p: RawDurationCount) => ({
        duration: Duration.fromNano(p.duration),
        count: p.count,
    });

    const createHistogram = (rawHistogram: Uint8Array) => {
        // wasm function.
        const miniHistogram = deserializeHistogram(rawHistogram, width);
        const percentiles = miniHistogram.percentiles.map(toDurationPercentile);
        const histogram = miniHistogram.data.map(toDurationCount);

        return {
            percentiles,
            histogram,
            max: Duration.fromNano(miniHistogram.max),
            min: Duration.fromNano(miniHistogram.min),
        };
    };

    const pollTimes =
        details.pollTimesHistogram.case === "histogram"
            ? createHistogram(details.pollTimesHistogram.value.rawHistogram)
            : // legacyHistogram
              createHistogram(details.pollTimesHistogram.value!);

    const scheduledTimes = details.scheduledTimesHistogram?.rawHistogram
        ? createHistogram(details.scheduledTimesHistogram.rawHistogram)
        : undefined;

    return {
        pollTimes,
        scheduledTimes,
    };
}
