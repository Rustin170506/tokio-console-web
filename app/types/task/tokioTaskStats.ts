import { Duration, Timestamp } from "../common/duration";
import type { Stats as ProtoTaskStats } from "~/gen/tasks_pb";

export interface TokioTaskStats {
    polls: bigint;
    createdAt: Timestamp;
    droppedAt?: Timestamp;
    busy: Duration;
    scheduled: Duration;
    lastPollStarted?: Timestamp;
    lastPollEnded?: Timestamp;
    idle?: Duration;
    total?: Duration;

    // === waker stats ===
    // Total number of times the task has been woken over its lifetime.
    wakes: bigint;
    // Total number of times the task's waker has been cloned
    wakerClones: bigint;
    // Total number of times the task's waker has been dropped.
    wakerDrops: bigint;
    // The timestamp of when the task was last woken.
    lastWake?: Timestamp;
    // Total number of times the task has woken itself.
    selfWakes: bigint;
}

export function fromProtoTaskStats(stats: ProtoTaskStats): TokioTaskStats {
    const createdAt = new Timestamp(
        stats.createdAt!.seconds,
        stats.createdAt!.nanos,
    );
    const droppedAt = stats.droppedAt
        ? new Timestamp(stats.droppedAt.seconds, stats.droppedAt.nanos)
        : undefined;
    const total = droppedAt?.subtract(createdAt);

    const pollStats = stats.pollStats!;
    const busy = new Duration(
        pollStats.busyTime?.seconds || 0n,
        pollStats.busyTime?.nanos || 0,
    );
    const scheduled = new Duration(
        stats.scheduledTime?.seconds || 0n,
        stats.scheduledTime?.nanos || 0,
    );
    const idle = total?.subtract(busy).subtract(scheduled);

    return {
        polls: pollStats.polls,
        createdAt,
        droppedAt,
        busy,
        scheduled,
        lastPollStarted: pollStats.lastPollStarted
            ? new Timestamp(
                  pollStats.lastPollStarted.seconds,
                  pollStats.lastPollStarted.nanos,
              )
            : undefined,
        lastPollEnded: pollStats.lastPollEnded
            ? new Timestamp(
                  pollStats.lastPollEnded.seconds,
                  pollStats.lastPollEnded.nanos,
              )
            : undefined,
        idle,
        total,
        wakes: stats.wakes,
        wakerClones: stats.wakerClones,
        wakerDrops: stats.wakerDrops,
        lastWake: stats.lastWake
            ? new Timestamp(stats.lastWake.seconds, stats.lastWake.nanos)
            : undefined,
        selfWakes: stats.selfWakes,
    };
}
