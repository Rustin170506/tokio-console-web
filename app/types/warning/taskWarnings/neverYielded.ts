import { TaskState, TokioTask } from "../../task/tokioTask";
import { Duration, Timestamp } from "../../common/duration";
import { type Warn, Warning } from "../warn";

const DEFAULT_DURATION: Duration = new Duration(1n, 0);

// Warning for if a task has never yielded
export class NeverYielded implements Warn<TokioTask> {
    name: string;
    minDuration: Duration;
    description: string;

    constructor(minDuration: Duration = DEFAULT_DURATION) {
        this.name = "NeverYielded";
        this.minDuration = minDuration;
        this.description = `tasks have never yielded (threshold ${minDuration.toString()})`;
    }

    check(task: TokioTask): Warning {
        // Don't fire warning for tasks that are not async.
        if (task.isBlocking()) {
            return Warning.Ok;
        }
        // Don't fire warning for tasks that are waiting to run.
        if (task.state() !== TaskState.Running) {
            return Warning.Ok;
        }

        if (task.totalPolls() > 1n) {
            return Warning.Ok;
        }

        // Avoid short-lived task false positives
        const busy = task.busyDuration(Timestamp.now());
        if (busy.greaterThanOrEqual(this.minDuration)) {
            return Warning.Warn;
        } else {
            return Warning.Recheck;
        }
    }

    format(task: TokioTask): string {
        const busy = task.busyDuration(Timestamp.now()).toString();
        return `This task has never yielded (${busy})`;
    }

    summary(): string {
        return this.description;
    }
}
