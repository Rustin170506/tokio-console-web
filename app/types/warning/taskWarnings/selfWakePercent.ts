import { Warning, type Warn } from "../warn";
import { TokioTask } from "../../task/tokioTask";

const DEFAULT_PERCENT = 50;

export class SelfWakePercent implements Warn<TokioTask> {
    name: string;
    minPercent: number;
    description: string;

    constructor(minPercent: number = DEFAULT_PERCENT) {
        this.name = "NeverYielded";
        this.minPercent = minPercent;
        this.description = `tasks have woken themselves over ${minPercent}% of the time`;
    }

    check(task: TokioTask): Warning {
        // Don't fire warning for tasks that are not async
        if (task.isBlocking()) {
            return Warning.Ok;
        }
        const selfWakes = task.selfWakePercent();
        if (selfWakes > this.minPercent) {
            return Warning.Warn;
        } else {
            return Warning.Ok;
        }
    }

    format(task: TokioTask): string {
        const selfWakes = task.selfWakePercent();
        return `This task has woken itself for more than ${this.minPercent}% of its total wakeups (${selfWakes}%)`;
    }

    summary(): string {
        return this.description;
    }
}
