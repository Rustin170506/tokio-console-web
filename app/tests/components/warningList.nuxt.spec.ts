import { expect, it } from "vitest";
import { mountSuspended, mockNuxtImport } from "@nuxt/test-utils/runtime";
import { createTask } from "./createTask";
import WarningList from "~/components/WarningList.vue";
import { TokioTask } from "~/types/task/tokioTask";
import { Timestamp } from "~/types/common/duration";
import { NeverYielded } from "~/types/warning/taskWarnings/neverYielded";

mockNuxtImport("useTasks", () => {
    return () => {
        const pending = ref<boolean>(false);
        const tasksData = ref<Map<bigint, TokioTask>>(new Map());
        const task = createTask();
        task.warnings.push(new NeverYielded());
        tasksData.value.set(1n, task);
        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));
        return { pending, tasksData, lastUpdatedAt };
    };
});

it("WarningsSlideover View", async () => {
    const component = await mountSuspended(WarningList);
    expect(component.html()).toMatchSnapshot();
});
