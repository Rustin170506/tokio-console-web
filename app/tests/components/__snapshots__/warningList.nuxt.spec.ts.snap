// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WarningList View 1`] = `
"<div class="p-4 flex-1 space-y-4">
  <div class="w-full relative overflow-hidden rounded-lg p-4 text-yellow-500 dark:text-yellow-400 ring-1 ring-inset ring-yellow-500 dark:ring-yellow-400">
    <div class="flex gap-3 items-start"><span class="i-heroicons-exclamation-triangle-20-solid flex-shrink-0 w-5 h-5"></span>
      <!--v-if-->
      <div class="w-0 flex-1">
        <p class="text-sm font-medium">NeverYielded</p>
        <div class="text-sm opacity-90 mt-0 leading-5">
          <p>1 tasks have never yielded (threshold 1.00s)</p>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </div>
  <div class="w-full relative overflow-hidden rounded-lg p-4 text-yellow-500 dark:text-yellow-400 ring-1 ring-inset ring-yellow-500 dark:ring-yellow-400">
    <div class="flex gap-3 items-start"><span class="i-heroicons-exclamation-triangle-20-solid flex-shrink-0 w-5 h-5"></span>
      <!--v-if-->
      <div class="w-0 flex-1">
        <p class="text-sm font-medium">LostWaker</p>
        <div class="text-sm opacity-90 mt-0 leading-5">
          <p>1 tasks have lost their wakers</p>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </div>
  <div class="w-full relative overflow-hidden rounded-lg p-4 text-yellow-500 dark:text-yellow-400 ring-1 ring-inset ring-yellow-500 dark:ring-yellow-400">
    <div class="flex gap-3 items-start"><span class="i-heroicons-exclamation-triangle-20-solid flex-shrink-0 w-5 h-5"></span>
      <!--v-if-->
      <div class="w-0 flex-1">
        <p class="text-sm font-medium">SelfWakePercent</p>
        <div class="text-sm opacity-90 mt-0 leading-5">
          <p>1 tasks have woken themselves over 50% of the time</p>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
